### DISCUSSION TITLE
Merging the chatgpt and whisper apis to be able to talk to gpt-powered NPCs.

### DISCUSSION THREAD

**Discussion by achraf-oujjir** [srcnalt/OpenAI-Unity#48] [Q&A] [answered]
Hello. I hope you are well.

I loved your video and your package. Easy to work with and minimal bugs.

I am very novice to unity and c# (1 month) and I have a group school project around "Education and VR". I have been tasked to add the feature of a talking npc (a professor) for which we'll make customized prompts. Different professors, different prompts for different subjects. A kind of virtual school. The idea came from Mesut's video on youtube: https://www.youtube.com/watch?v=QMogW-A9oD8&t=6s&ab_channel=MesutTomruk and I want to reproduce a similar interaction with the NPCs.
I want to know how I can connect the two APIs in regard to the following issues:

1) Is it possible to solve the problem of the duration of the recording? Having a conversation-like flow of the messages will be a very good thing to have. I am thinking maybe a constant recording that stops when the person stays silent for more than 3 seconds. It might cost in RAM and CPU which will make the replies even slower. Is there another solution ?

2) Getting very long prompts. While using chatgpt, it seems like it doesn't finish its answers sometimes. Is there a way to bypass that ??

Another issue is, if I want to interrupt the npc mid-sentence.

Thank you in advance for your help.

**Comment by srcnalt** [comment, marked as answer]
Hi @achraf-oujjir,

- Yes, you can make the audio sent longer, in the sample it was 5 seconds however OpenAI does not show a max amount in the API Docs. You can turn the sample code into a push-to-talk system which should work for you.

- You can add a part to your prompt such as "Reply in maximum 100 words" then this should help you shorten the result.

**Reply by achraf-oujjir** [reply]
For the limit of whisper, I believe I saw somewhere something about the max number of tokens. I have thought about the push-to-talk solution and I am implementing it now. However, in line 46 of Whisper.cs, I don't know how to modify the line:

> clip = Microphone.Start(dropdown.options[index].text, false, duration, 44100);

I am reading the documentation of the method Microphone.Start and I am trying to make the duration variable. It shouldn't stop recording until the user pushes 'Stop' button. Is there a way to do it ??

**Reply by achraf-oujjir** [reply]
I think I'll make duration a very big number in seconds (ex: 300, which is 5 minutes) and then put in an event to end the recording when the 'Stop' button is pushed. Is this a good way of implementing it ? If not, kindly enlighten me.

**Reply by achraf-oujjir** [reply]
Update: I was able to marry the two APIs, chatgpt and whisper.

However, I have an issue with whisper specifically. For example, I said: "What is unity ?"

Here is the Debug.Log of the text returned by whisper:
`This is res.Text: What is unity? Unity is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind. The mind is the power of the mind.
UnityEngine.Debug:Log (object)
OpenAI.ProfGPT/<SaveTranscribeSend>d__15:MoveNext () (at Assets/Samples/OpenAI Unity/0.1.12/ProfGPT/ProfGPT.cs:92)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<OpenAI.CreateAudioResponse>:SetResult (OpenAI.CreateAudioResponse)
OpenAI.OpenAIApi/<CreateAudioTranscription>d__21:MoveNext () (at Library/PackageCache/com.srcnalt.openai-unity@fb8af0a6f6/Runtime/OpenAIApi.cs:373)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<OpenAI.CreateAudioResponse>:SetResult (OpenAI.CreateAudioResponse)
OpenAI.OpenAIApi/<DispatchRequest>d__8`1<OpenAI.CreateAudioResponse>:MoveNext () (at Library/PackageCache/com.srcnalt.openai-unity@fb8af0a6f6/Runtime/OpenAIApi.cs:182)
UnityEngine.UnitySynchronizationContext:ExecuteTasks ()
`
ChatGPT is confused:

> The returned message is: I'm sorry, but your question seems incomplete. Are you asking for my definition of unity in a specific context? Unity can refer to the state of being united or combined into one whole, or it can refer to a sense of solidarity or shared purpose among a group of people. It can mean different things depending on the context in which it is used. Can you provide me with more information or a specific question so I can provide a more relevant answer for you?

Is there a solution for the whisper problem ??

**Reply by srcnalt** [reply]
This is what OpenAI endpoints produce there is really nothing I can do about what comes from that side.

**Reply by achraf-oujjir** [reply]
What's a bit weird is that, it works fine in the Whisper sample project that you provided in your github repo. The transcribed text is exactly what I am saying. Maybe it's something from my part. I will attach here the script (in .txt) that I wrote that joins the two APIs for the goal of making a gpt-powered npc professor. If you want to reproduce it, I have made a sample scene with a canvas containing a dropdown, two buttons (record and stop) and an image that will show the transcribed text.

[ProfGPT.txt](https://github.com/srcnalt/OpenAI-Unity/files/11343194/ProfGPT.txt)

**Reply by achraf-oujjir** [reply]
It was the duration of the clip. When I got it down to 10s or even 30s, it works fine.
Now to the next step, text to speech to hear the answer of GPT.

I saw your video about text to speech using AWS Polly service. Do you have other (preferably free) APIs for text to speech ??

I have found Resemble AI and Google's API. I still don't know how to integrate them within unity. I found a couple github repos but I still need to check if it integrates well with my app and with the 3d character (as I'll need to do lip sync later)

**Reply by achraf-oujjir** [reply]
Update: I worked with AWS Polly service and was able to merge it with whisper and chatgpt! Now to the oculus lip sync

**Reply by achraf-oujjir** [reply]
Update: I finished the ProfGPT character with Lip Sync and pronounciation in english and french (with french accent). I liked your tutorial, very clear and straight to the point even for me who's a novice to Unity and C#! Thank you @srcnalt for your youtube videos and your help here on github. Next step for me is to include ProfGPT in different scenes for different subjects.

Looking forward to your next video on youtube!

**Reply by GamedevBranislav** [reply]
Hi i been recently trying to make working chatgpt in Unity, but now it says is doesn't work for free anymore and it is for tokens(credit card) did you make this without spending money or there is different way around instead of using openai? thanks

**Reply by achraf-oujjir** [reply]
Hey @GamedevBranislav. Where did you find that it's not free anymore ?? Did you get your openai api keys and test the chatgpt sample scene in this github repo ??

**Reply by GamedevBranislav** [reply]
Yes I use mine keys I used openai and unity just print error : "You exceeded your current quota, please check your plan and billing details" 
Yours working just fine?

**Reply by achraf-oujjir** [reply]
I haven't tested it in a week or so. Did you use the api extensively or did you just get your api keys ? Are they yours ? Have you shared them with other people ??

**Reply by GamedevBranislav** [reply]
I used only my keys I generate new, I didn't share with anyone, maybe I did something wrong but I wanna confirm this, it is possible make game build and which works so people can download and play with?

**Reply by achraf-oujjir** [reply]
I just tested right now and I get the same message as you. Here is the link for the api rates: https://platform.openai.com/docs/guides/rate-limits/overview and here is a stackoverflow post about it: https://stackoverflow.com/questions/75898276/openai-chatgpt-gpt-3-5-api-error-429-you-exceeded-your-current-quota-please

**Reply by achraf-oujjir** [reply]
Hey @GamedevBranislav, did you solve your issue ? if yes, how ??

**Comment by zicas2000** [comment]
@achraf-oujjir 
If I may allowed to contribute in this discussion, this is a good solution to text to speech ,it's free and easy to implement,
https://www.youtube.com/watch?v=XRXbVtr1fog

**Reply by achraf-oujjir** [reply]
Interesting! I should check if it gives faster speech.
